pkg/mcs/resourcemanager/server/keyspace_manager.go

```diff
@@ -346,3 +347,185 @@ func (rt *ruTracker) getRUPerSec() float64 {
+			for _, group := range queue {
+				ruTracker := krgm.getRUTracker(group.Name)
+				if ruTracker == nil || ruTracker.getRUPerSec() == 0 {
+					continue
+				}
+				// Only set the active resource groups to avoid unnecessary overrides.
+				group.overrideFillRateAndBurstLimit(0, 0)
+			}
+			continue
+		}
+
+		var (
+			totalBasicRUDemand = 0.0
+			totalBurstRUDemand = 0.0
+			basicRUDemandMap   = make(map[string]float64, len(queue))
+			burstRUDemandMap   = make(map[string]float64, len(queue))
+		)
+		// Calculate the basic and burst RU demands for all groups in this priority level.
+		for _, group := range queue {
+			ruTracker := krgm.getRUTracker(group.Name)
+			// Not found the RU tracker, skip this group.
+			if ruTracker == nil {
+				continue
+			}
+			ruPerSec := ruTracker.getRUPerSec()
+			fillRateSetting := group.getFillRateSetting()
+			burstLimitSetting := group.getBurstLimitSetting()
+			// Calculate the basic RU demand of the resource group.
+			// Basic demand is the minimum of real-time RU consumption and configured fill rate.
+			basicRUDemand := math.Min(ruPerSec, fillRateSetting)
+			totalBasicRUDemand += basicRUDemand
+			basicRUDemandMap[group.Name] = basicRUDemand
+			// Calculate the burst RU demand of the resource group.
+			// Burst demand is the excess consumption beyond the fill rate setting.
+			burstRUDemand := math.Max(0, ruPerSec-fillRateSetting)
+			// If the burst limit setting is greater than 0, the burst RU demand should not exceed the burst limit setting.
+			if burstLimitSetting >= 0 {
+				burstRUDemand = math.Min(burstRUDemand, burstLimitSetting)
+			}
+			totalBurstRUDemand += burstRUDemand
+			burstRUDemandMap[group.Name] = burstRUDemand
+		}
+
+		totalRUDemand := totalBasicRUDemand + totalBurstRUDemand
+		// If the total real-time RU demand is greater than or equal to the remaining service limit,
+		// we need to divide the remaining service limit proportionally within the same priority level.
+		if totalRUDemand >= remainingServiceLimit {
+			// If the basic RU demand already exceeds the remaining service limit, then we can only try to meet the basic needs of
+			// the resource groups through proportional allocation of the fill rate setting.
+			if totalBasicRUDemand >= remainingServiceLimit {
+				for _, group := range queue {
+					basicRUDemand := basicRUDemandMap[group.Name]
+					// Allocate the remaining service limit proportionally based on basic demand.
+					overrideFillRate := math.Min(
+						remainingServiceLimit*basicRUDemand/totalBasicRUDemand,
+						group.getFillRateSetting(), // Should not exceed the original fill rate setting.
+					)
+					// Do not allow the resource group to consume extra tokens in this case,
+					// so the override burst limit is set to the same as the override fill rate.
+					group.overrideFillRateAndBurstLimit(overrideFillRate, int64(overrideFillRate))
+				}
+				// After allocation, all remaining service limit is consumed.
+				remainingServiceLimit = 0
```

Comment:
If `group.getFillRateSetting()` is less than `remainingServiceLimit*basicRUDemand/totalBasicRUDemand`, it might still have `remainingServiceLimit`?